### DISCUSSION TITLE
Default `timestamptz` precision in Postgres is too high, causes queries (and concurrency checks) to fail.

### DISCUSSION THREAD

**Discussion by CynicalBusiness** [mikro-orm/mikro-orm#6019] [General]
### Describe the bug

As noted in the [docs](https://mikro-orm.io/docs/upgrading-v5-to-v6#changed-default-postgresql-date-mapping-precision), the default `timestamptz` precision for postgresql is simply "not specified", causing postgres to assume `timestamptz(6)`. The JavaScript `Date`, however, stores only up to 3 points of precision; ordinarily this doesn't present an issue as postgres just fills the remaining three with zeros.

However, I got stuck for a few hours trying to diagnose `OptimisticConcurrencyError` (I'm using a `Date` column for this) only to discover it appears to be an ORM/Postgres issue.

More specifically, I have a column configured as such:
```typescript
@Property({
    version: true,
})
updatedAt!: Date;
```

When this entity is inserted, it correctly fills out, for example, `2024-09-10 11:42:01.345 +0200` in the column. However, when I try to update this entity later, the concurrency check fails because `WHERE updatedAt = '2024-09-10 11:42:01.345 +0200'` returns zero rows, even though the date is correct in the DB.

As it turns out, postgres had encountered a floating-point error (?) when interpreting the date, and *actually* stored the date as `2024-09-10 11:42:01.345001 +0200`.

Mikro has been working flawlessly with this on other entities up until this point, just this one particular entity with this one particular date experienced the problem. I have an extremely small dataset of only a dozen-or-so rows in this case, so I imagine this issue *must* present itself at least occasionally in larger datasets, causing silent failures.

I was able to get it working again by manually changing the column with `length: 3`, resulting in a `timestamptz(3)` column type. I cannot, however, find any information on how to change this globally for all entites/dates, meaning I have to remember to do it for every date type lest this issue arise again.

### Reproduction

Reproduction of this is extremely finicky and seemingly up to chance, as it is an issue with precision within the Postgres engine. It's not an ORM *error* per se, but rather a misconfigured default setting invoking an edge case.

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.3.3

### Node.js version

20.10.0

### Operating system

Windows 11

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
Using dates for version columns is a bad idea if you ask me, exactly for this reason. You can get conflicts even with the 6-digit precision, which is less likely, but the problem remains.

You can alter metadata globally in several ways:

https://mikro-orm.io/docs/configuration#adjusting-default-type-mapping
https://mikro-orm.io/docs/configuration#onmetadata-hook

**Reply by CynicalBusiness** [reply]
> You can alter metadata globally in several ways:
> 
> https://mikro-orm.io/docs/configuration#adjusting-default-type-mapping
> https://mikro-orm.io/docs/configuration#onmetadata-hook

*This* is what I was looking for to set things globally, thanks! Wasn't quite sure what it was called and just couldn't find it myself.

> Using dates for version columns is a bad idea if you ask me, exactly for this reason. You can get conflicts even with the 6-digit precision, which is less likely, but the problem remains.

Regarding the whole version-column thing: the expected concurrency on an individual entity in this app is quite low, and the `updatedAt` column already existed for auditing purposes, so it seemed reasonable.

The precision problem, though, appears in non-versioned columns. I just happened to notice it because the error changes from a silent fail to making itself known with a conflict error.
